python-oauth-web-app/app.py

```diff
@@ -0,0 +1,336 @@
+    # DID was supplied for login.
+    username = request.form["username"]
+    if not username.startswith("did:"):
+        if not valid_handle(username):
+            flash("Invalid Handle")
+            return render_template("login.html"), 400
+        did = resolve_handle(username)
+        if not did:
+            flash("Handle Not Found")
+            return render_template("login.html"), 400
+    else:
+        did = username
+    if not valid_did(did):
+        flash("Invalid DID: " + did)
+        return render_template("login.html"), 400
+    did_doc = resolve_did(did)
+    if not did_doc:
+        flash("DID Not Found: " + did)
+        return render_template("login.html"), 400
+
+    # Fetch PDS OAuth metadata.
+    # IMPORTANT: PDS endpoint URL is untrusted input, SSRF mitigations are needed
+    pds_url = pds_endpoint(did_doc)
+    print(f"account PDS: {pds_url}")
+    assert is_safe_url(pds_url)
+    with hardened_http.get_session() as sess:
+        resp = sess.get(f"{pds_url}/.well-known/oauth-protected-resource")
+    resp.raise_for_status()
+    # Additionally check that status is exactly 200 (not just 2xx)
+    assert resp.status_code == 200
+
+    # Fetch Auth Server metadata. For a self-hosted PDS, this will be the same server (the PDS). For large-scale PDS hosts like Bluesky, this may be a separate "entryway" server filling the Auth Server role.
+    # IMPORTANT: Authorization Server URL is untrusted input, SSRF mitigations are needed
+    authsrv_url = resp.json()["authorization_servers"][0]
+    print(f"account Auth Server: {authsrv_url}")
+    assert is_safe_url(authsrv_url)
+    try:
+        authsrv_meta = fetch_authsrv_meta(authsrv_url)
+    except Exception as err:
+        print(f"failed to fetch auth server metadata: " + str(err))
+        flash("Failed to fetch Auth Server (Entryway) OAuth metadata")
+        return render_template("login.html"), 400
+
+    # Dynamically compute our "client_id" based on the request HTTP Host
+    app_url = request.url_root.replace("http://", "https://")
+    client_id = f"{app_url}oauth/client-metadata.json"
+
+    # Generate DPoP private signing key for this account session. In theory this could be defered until the token request at the end of the athentication flow, but doing it now allows early binding during the PAR request.
+    # TODO: nonce?
+    dpop_private_jwk = JsonWebKey.generate_key("EC", "P-256", is_private=True).as_json(
+        is_private=True
+    )
+
+    # Submit OAuth Pushed Authentication Request (PAR). We could have constructed a more complex authentication request URL below instead, but there are some advantages with PAR, including failing fast, early DPoP binding, and no URL length limitations.
+    # TODO: use dpop_private_jwk here, along with a nonce?
+    pkce_verifier, state, nonce, resp = send_par_auth_request(
+        authsrv_url, authsrv_meta, did, app_url, secret_jwk
```

Comment:
Right now this comes through into the login screen without being re-resolved back to the handle which is quite confusing:

<img width="951" alt="image" src="https://github.com/user-attachments/assets/5dac5639-a1c6-4fc9-8dbd-950a414c2d3b">

Is it possible to send the handle here instead of `did`?  If not, can the oauth frontend to resolve this to show the handle to the user?

Reply:
@bnewbold you can use the user provided handle as `login_hint` instead of the did.

Note that the oauth-provider implementation should probably handle this better too.

Reply:
updated to make this actually work